File: packages/cli/hieudoanm/cobra.go/cmd/clipboard/clipboard.go
package clipboard

import (
	"database/sql"
	"fmt"
	"log"
	"time"

	"github.com/atotto/clipboard"
	_ "github.com/mattn/go-sqlite3"
	"github.com/spf13/cobra"
)

func NewCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "clipboard",
		Short: "Watch clipboard changes and store them in SQLite",
		Long:  `Monitors the system clipboard for changes and saves each unique entry to a local SQLite database.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			runClipboardWatcher()
			return nil
		},
	}
}

func runClipboardWatcher() {
	db, err := sql.Open("sqlite3", "~/.hieudoanm/hieudoanm.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	createTable(db)

	fmt.Println("Clipboard watcher started... (Ctrl+C to stop)")

	var lastText string

	for {
		text, err := clipboard.ReadAll()
		if err != nil {
			time.Sleep(1 * time.Second)
			continue
		}

		if text == "" || text == lastText {
			time.Sleep(500 * time.Millisecond)
			continue
		}

		err = insertClip(db, text)
		if err != nil {
			log.Println("insert error:", err)
		} else {
			fmt.Println("Saved:", preview(text))
		}

		lastText = text
		time.Sleep(500 * time.Millisecond)
	}
}

func createTable(db *sql.DB) {
	query := `
	CREATE TABLE IF NOT EXISTS clips (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		content TEXT NOT NULL,
		created_at DATETIME DEFAULT CURRENT_TIMESTAMP
	);`
	_, err := db.Exec(query)
	if err != nil {
		log.Fatal(err)
	}
}

func insertClip(db *sql.DB, content string) error {
	_, err := db.Exec(
		`INSERT INTO clips (content) VALUES (?)`,
		content,
	)
	return err
}

func preview(text string) string {
	if len(text) > 40 {
		return text[:40] + "..."
	}
	return text
}

File: packages/cli/hieudoanm/clap.rs/src/cmd/system/clipboard.rs
use std::time::Duration;

#[derive(clap::Args)]
pub struct Args;

pub fn command() -> clap::Command {
    clap::Command::new("clipboard").about("Listen to clipboard changes and store them")
}

fn preview_text(text: &str) -> String {
    if text.len() > 40 {
        format!("{}...", &text[..40])
    } else {
        text.to_string()
    }
}

pub async fn run(_matches: &Args) -> anyhow::Result<()> {
    use rusqlite::Connection;

    let home = dirs::home_dir().ok_or_else(|| anyhow::anyhow!("cannot find home dir"))?;
    let dir = home.join(".hieudoanm");
    std::fs::create_dir_all(&dir)?;
    let path = dir.join("hieudoanm.db");
    let conn = Connection::open(&path)?;
    conn.execute_batch("CREATE TABLE IF NOT EXISTS clips (id INTEGER PRIMARY KEY AUTOINCREMENT, content TEXT NOT NULL, created_at DATETIME DEFAULT CURRENT_TIMESTAMP);")?;

    #[cfg(target_os = "macos")]
    let read_clip = || -> anyhow::Result<String> {
        let output = std::process::Command::new("pbpaste")
            .output()
            .map_err(|e| anyhow::anyhow!("pbpaste failed: {e}"))?;
        Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
    };

    #[cfg(not(target_os = "macos"))]
    let read_clip =
        || -> anyhow::Result<String> { anyhow::bail!("clipboard not supported on this platform") };

    println!("Clipboard watcher started... (Ctrl+C to stop)");
    let mut last_text = String::new();
    loop {
        if let Ok(text) = read_clip() {
            if !text.is_empty() && text != last_text {
                conn.execute(
                    "INSERT INTO clips (content) VALUES (?1)",
                    rusqlite::params![text],
                )?;
                let preview = preview_text(&text);
                println!("Saved: {preview}");
                last_text = text;
            }
        }
        std::thread::sleep(Duration::from_millis(500));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_command_definition() {
        let cmd = command();
        assert_eq!(cmd.get_name(), "clipboard");
    }

    #[test]
    fn test_preview_text_short() {
        assert_eq!(preview_text("hello"), "hello");
    }

    #[test]
    fn test_preview_text_exactly_40() {
        let s = "a".repeat(40);
        assert_eq!(preview_text(&s), s);
    }

    #[test]
    fn test_preview_text_long() {
        let s = "a".repeat(100);
        let expected = format!("{}...", "a".repeat(40));
        assert_eq!(preview_text(&s), expected);
        assert_eq!(preview_text(&s).len(), 43);
    }

    #[test]
    fn test_preview_text_empty() {
        assert_eq!(preview_text(""), "");
    }

    #[test]
    fn test_preview_text_unicode() {
        let s = "🦀".repeat(30);
        assert_eq!(preview_text(&s).len(), 43);
        assert!(preview_text(&s).ends_with("..."));
    }
}
